### PR TITLE
[c++] `can_upgrade_domain`

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** As tracked on issue #2407 / [[sc-51048]](https://app.shortcut.com/tiledb-inc/story/51048).

Note that the intended Python and R API changes are all agreed on and finalized as described in #2407.

**Changes:**

* This is `can_upgrade_domain` for `SOMADataFrame` at the C++ level

**Notes for Reviewer:**

### CODE REVIEW

File: libtiledbsoma/src/soma/soma_array.cc [hunk 1]
```diff
@@ -1882,6 +1882,189 @@ void SOMAArray::_set_soma_joinid_shape_helper(
     schema_evolution.array_evolve(uri_);
 }
 
+std::pair<bool, std::string> SOMAArray::can_upgrade_domain(
```
Comment (nguyenv) [inline libtiledbsoma/src/soma/soma_array.cc:1885]: Did we add a `using StatusAndReason = std::pair<bool, std::string>` alias in a previous PR? I thought we did but now can't find it.

File: libtiledbsoma/src/soma/soma_array.h [hunk 2]
```diff
@@ -1157,6 +1157,12 @@ class SOMAArray : public SOMAObject {
             newshape, false, function_name_for_messages);
     }
 
+    /**
+     * This is for SOMADataFrame.
+     */
+    std::pair<bool, std::string> can_upgrade_domain(
+        const ArrowTable& newdomain, std::string function_name_for_messages);
```
Comment (nguyenv) [inline libtiledbsoma/src/soma/soma_array.h:1164]: Can this just reside in the `SOMADataFrame` class or are there issues with doing that, so we need to place it in `SOMAArray`?

**Review by nguyenv — APPROVED** [review]

File: libtiledbsoma/src/soma/soma_array.cc (hunk 1 above)
Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.cc:1885]: oh yes -- indeed, let me check 👀

File: libtiledbsoma/src/soma/soma_array.h (hunk 2 above)
Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.h:1164]: There are issues with doing that:
https://github.com/single-cell-data/TileDB-SOMA/blob/62dd129515c8e6928d910a2d20f6e0ac431c9fe7/libtiledbsoma/src/soma/soma_array.h#L1230-L1234

I can propagate that comment material

Comment (nguyenv) [reply libtiledbsoma/src/soma/soma_array.h:1164]: Ah ic- thanks. No need to copy it over if it's already in the header.

Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.h:1164]: Adding as top-of-file comment :)

File: libtiledbsoma/src/soma/soma_array.cc (hunk 1 above)
Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.cc:1885]: I was puzzled too! Turns out the reason neither of us could find this is it was done in Python not C++ :D 
https://github.com/single-cell-data/TileDB-SOMA/blob/62dd129515c8e6928d910a2d20f6e0ac431c9fe7/apis/python/src/tiledbsoma/_types.py#L82-L84

It's worth adapting here too though.

Comment (johnkerl) [reply libtiledbsoma/src/soma/soma_array.cc]: (added)

**Pull request merged**